Userland/Libraries/LibGfx/PNGWriter.cpp

```diff
@@ -0,0 +1,195 @@
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "PNGWriter.h"
+#include <AK/String.h>
+#include <LibCrypto/Checksum/CRC32.h>
+#include <stdio.h>
+
+namespace Gfx {
+
+PNGChunk::PNGChunk(const String& type)
+    : m_type(move(type))
+{
+}
+
+void PNGChunk::add_u8(u8 data)
+{
+    m_data.append(data);
+}
+
+void PNGChunk::add_u16_little(u16 data)
+{
+    m_data.append(data & 0xff);
+    m_data.append((data >> 8) & 0xff);
+}
+
+void PNGChunk::add_u32_little(u32 data)
+{
+    m_data.append(data & 0xff);
+    m_data.append((data >> 8) & 0xff);
+    m_data.append((data >> 16) & 0xff);
+    m_data.append((data >> 24) & 0xff);
+}
+
+void PNGChunk::add_u32_big(u32 data)
+{
+    m_data.append((data >> 24) & 0xff);
+    m_data.append((data >> 16) & 0xff);
+    m_data.append((data >> 8) & 0xff);
+    m_data.append(data & 0xff);
+}
+
+void PNGChunk::add_u16_big(u16 data)
+{
+    m_data.append((data >> 8) & 0xff);
+    m_data.append(data & 0xff);
+}
+
+void NonCompressibleBlock::add_byte_to_block(u8 data, PNGChunk& chunk)
+{
+    m_non_compressible_data.append(data);
+    update_adler(data);
+    if (full()) {
+        add_block_to_chunk(chunk, false);
+        m_non_compressible_data.clear();
+    }
+}
+
+void NonCompressibleBlock::add_block_to_chunk(PNGChunk& png_chunk, bool last)
+{
+    if (last) {
+        png_chunk.add_u8(1);
+    } else {
+        png_chunk.add_u8(0);
+    }
+
+    auto len = m_non_compressible_data.size();
+    auto nlen = ~len;
+
+    png_chunk.add_u16_little(len);
+    png_chunk.add_u16_little(nlen);
+
+    for (auto non_compressed_byte : m_non_compressible_data) {
+        png_chunk.add_u8(non_compressed_byte);
+    }
+}
+
+void NonCompressibleBlock::finalize(PNGChunk& chunk)
+{
+    add_block_to_chunk(chunk, true);
+}
+
+void NonCompressibleBlock::update_adler(u8 data)
+{
+    m_adler_s1 = (m_adler_s1 + data) % 65521;
+    m_adler_s2 = (m_adler_s2 + m_adler_s1) % 65521;
+}
+
+void PNGWriter::add_chunk(PNGChunk& png_chunk)
```

Comment:
Seems like this should take a `const PNGChunk&` since it's not mutating it.

Reply:
Done.